### PR TITLE
Integration Specs for Views and Fix n+1 Problems

### PR DESCRIPTION
**Pull request by iamchristianani**
## Integration Specs for Views and Fix n+1 Problems for the Blog App

> In this Pull Request, @MohamedHNoor and @iamchristianani completed the following:

- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user.
  - Before the fix
n+1 before

  - After the fix
n+1 after

- Used Capybara to write integration tests for each view in our project.
- Wrote different tests for the following Pages:
  - User Index Page
  - User Show Page
  - User Post Index Page
  - Post Show Page

### CODE REVIEW

File: spec/views/posts/index.html.erb_spec.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+  it 'shows the number of likes on the post' do
+    expect(page).to have_content('Likes: 1')
+  end
+
+  it "when clicking on a post title, it redirects to the post's show page" do
+    click_on @first_post.title
+    expect(page).to have_content(@first_post.title)
+  end
+end
```
Comment (alexander16108) [inline spec/views/posts/index.html.erb_spec.rb]: please add the last test required for the post index 😄 

`Tests that asserts you can see a section for pagination if there are more posts than fit on the view.`

**Review by alexander16108 — CHANGES_REQUESTED** [review]
Hi @iamchristianani @MohamedHNoor , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

image


## Highlights
All linter checks are passing ✔️ ✔️
Your PR is professional 👏 👏

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/posts/index.html.erb_spec.rb (hunk 1 above)
Comment (MohamedHNoor) [reply spec/views/posts/index.html.erb_spec.rb]: Thanks, it's done

**Review by alexander16108 — APPROVED** [review]
Hi @iamchristianani @MohamedHNoor , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

image


## Highlights
All linter checks are passing ✔️ ✔️
Your PR is professional 👏 👏

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by nicupop729** [comment]
Hi Team,

The issue was already addressed, and the submission on @MohamedHNoor was invalidated in the dashboard. @MohamedHNoor please use your own PR link to ask for a review.

@iamchristianani, the approval of @alexander16108 is not taken into consideration. It was supposed to be invalidated as your partner wrongly used your PR to ask for a review. So @iamchristianani, please use your PR now to ask again for another review.

@next_code_reviewer, please disregard completely the approval summary of @alexander16108 and please be consistent with his first review, which was provided correctly, for @iamchristianani.

Good luck!

**Review by nicupop729 — COMMENTED** [review]
Hi Team,

The issue was already addressed, and the submission on @MohamedHNoor was invalidated in the dashboard. @MohamedHNoor please use your own PR link to ask for a review.

@iamchristianani, the approval of @alexander16108 is not taken into consideration. It was supposed to be invalidated as your partner wrongly used your PR to ask for a review. So @iamchristianani, please use your PR now to ask again for another review.

@next_code_reviewer, please disregard completely the approval summary of @alexander16108 and please be consistent with his first review, which was provided correctly, for @iamchristianani.

Good luck!

File: spec/views/posts/index.html.erb_spec.rb [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+  it "when clicking on a author's title, it redirects to the post's show page" do
+    click_on @first_post.title
+    expect(page).to have_content(@first_post.title)
+  end
+
+  it "when click on a post, it redirects to the post's show page" do
+    click_on @first_post.title
+    expect(page).to have_content(@first_post.title)
+  end
```
Comment (danmainah) [inline spec/views/posts/index.html.erb_spec.rb:56]: - [ ] @iamchristianani as previously requested kindly ensure to add a test for pagination, where users can see a section of pagination if all posts can fit the view.

**Review by danmainah — COMMENTED** [review]
Hi @iamchristianani , 

Great work on making the changes requested by a previous reviewer 👏  👏 👏
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
✔️ No linter errors
✔️ Gitflow was correctly used
✔️  Good design


 # Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**